### PR TITLE
Bookstore order

### PR DESCRIPTION
**Pull request by azzilian**
Order entity

Order: Represents an order placed by a user.

    Add a new entity Order with the next fields:
        id (Long, PK)
        user (User, not null)
        status (Status (enum), not null)
        total (BigDecimal, not null)
        orderDate (LocalDateTime, not null)
        shippingAddress (String, not null)
        orderItems (Set<OrderItem>)

OrderItem entity

OrderItem: Represents an item in a user's order.

    Add a new entity OrderItem with the next fields:
        id (Long, PK)
        order (Order, not null)
        book (Book, not null)
        quantity (int, not null)
        price (BigDecimal, not null)

### CODE REVIEW

File: src/main/java/com/onlinebookstore/onlinebookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        return orderService.getOrderHistory(user);
+    }
+
+    @GetMapping("/{orderId}/items")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get a all items in order", description = " User can view "
+            + "all items in order")
+    public Set<OrderItemResponseDto> getOrderItems(
+            @PathVariable Long orderId) {
+        return orderService.getOrderItems(orderId);
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/controller/OrderController.java]: Please make sure that one user can`t access order data of different user (check all endpoints)

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please, CHECK PR BEFORE PUSHING to review, and prepare PR wiht necessary changes only

**Comment by azzilian** [comment]
> Please, CHECK PR BEFORE PUSHING to review, and prepare PR wiht necessary changes only

A this is commits that come along after I updated my master

File: src/main/java/com/onlinebookstore/onlinebookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Create order", description = "Creating new order"
+            + " based on cartItems in users cart, after order created - "
+            + "all cartitems are removed from cart, status became PENDING")
+    public OrderResponseDto placeOrder(
+            @Valid @RequestBody OrderRequestDto orderRequestDto,
+            @AuthenticationPrincipal User user) {
+        orderRequestDto.setUserId(user.getId());
+        return orderService.placeOrder(orderRequestDto);
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/controller/OrderController.java]: order RequestDto represents information that user send to us, I suggest to not add there a user, but to pass it like another argument to 'place Order' method

File: src/main/java/com/onlinebookstore/onlinebookstore/dto/order/OrderItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.onlinebookstore.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class OrderItemRequestDto {
+    @Min(value = 1, message = "minimal quantity is at least 1")
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/dto/order/OrderItemRequestDto.java:10]: Wrong message

File: src/main/java/com/onlinebookstore/onlinebookstore/dto/order/OrderRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+@Data
+public class OrderRequestDto {
+    private Long userId;
+    @NotNull(message = "shippingAddress field cannot be empty")
+    @Size(min = 1, message = "shippingAddress should be at least 1 character")
+    private String shippingAddress;
+    private BigDecimal total;
+    private LocalDateTime orderDate;
+    private Set<OrderItemRequestDto> orderItemDto;
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/dto/order/OrderRequestDto.java]:
```suggestion
    @NotNull(message = "shippingAddress field cannot be empty")
    @Size(min = 1, message = "shippingAddress should be at least 1 character")
    private String shippingAddress;

```

File: src/main/java/com/onlinebookstore/onlinebookstore/model/orders/Order.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import java.time.LocalDateTime;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/model/orders/Order.java:30]: Add equals, hashCode, toString to all entities

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    private final UserRepository userRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final ShoppingCartService shoppingCartService;
+
+    @Override
+    public OrderResponseDto placeOrder(OrderRequestDto orderRequestDto) {
+        User user = userRepository.findById(orderRequestDto.getUserId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "User not found with id: " + orderRequestDto.getUserId()));
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: No reason to make DB request, you can take user from securityContext

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                .map(cartItem -> cartItem.getBook()
+                        .getPrice()
+                        .multiply(BigDecimal
+                                .valueOf(cartItem
+                                        .getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add));
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(orderRequestDto.getShippingAddress());
+
+        Order savedOrder = orderRepository.save(order);
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: Please move logic of order creation to separate method

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+        cart.getCartItems().forEach(cartItem -> shoppingCartService
+                .removeBookFromCart(cartItem.getId(), user.getId()));
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: You can delete all cartItems using one DB request, not reason to make N requests

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    @Override
+    public OrderResponseDto updateOrderStatus(Long orderId,
+                                              OrderUpdateStatusDto orderUpdateStatusDto) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order not found with id: "
+                        + orderId));
+        System.out.println("Updating order status to: "
+                + orderUpdateStatusDto.getOrderStatus());
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: If you want some additional info in console use Logger

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        order.setOrderStatus(orderUpdateStatusDto.getOrderStatus());
+
+        if (order.getOrderStatus() == OrderStatus.COMPLETED) {
+            ShoppingCart cart = shoppingCartRepository.findByUserId(order.getUser().getId())
+                    .orElseThrow(() -> new EntityNotFoundException(
+                            "Shopping cart not found for user id: "
+                                    + order.getUser().getId()));
+            cart.getCartItems().clear();
+            shoppingCartRepository.save(cart);
+        }
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: Why this logic required? You already cleaned shopping cart when order was created

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+                                                   Long userId) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Order not found with id: " + orderId));
+
+        Set<OrderItem> orderItems = order.getOrderItems();
+
+        if (!order.getUser().getId().equals(userId)) {
+            throw new SecurityException("You do not have permission to see this order items");
+        }
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java:91]: Code duplication

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left some comments

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java (hunk 9 above)
Comment (azzilian) [reply src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: It was drastic measure to find bug))

File: src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java (hunk 10 above)
Comment (azzilian) [reply src/main/java/com/onlinebookstore/onlinebookstore/service/impl/OrderServiceImpl.java]: Honestly I got some idea about this but right now cannot remeber it... I guess i will delete it then

**Pull request merged**